api.md

```diff
@@ -0,0 +1,215 @@
+
+This document provides a mix of:
+
+* **High level design guidance** that individual APIs interpret to meet their needs.
+* **Low level web practices** that most modern HTTP APIs use.
+
+### Design for common use cases
+
+For APIs that syndicate data, consider several common client use cases:
+
+* **Bulk data.** Clients often wish to establish their own copy of the API's dataset in its entirety. For example, someone might like to build their own search engine on top of the dataset, using different parameters and technology than the "official" API allows. If the API can't easily act as a bulk data provider, provide a separate mechanism for acquiring the backing dataset in bulk.
+* **Staying up to date.** Especially for large datasets, clients may want to keep their dataset up to date without downloading the data set after every change. If this is a use case for the API, prioritize it in the design.
+* **Driving expensive actions.** What would happen if a client wanted to automatically send text messages to thousands of people or light up the side of a skyscraper every time a new record appears? Consider whether the API's records will always be in a reliable unchanging order, and whether they tend to appear in clumps or in a steady stream. Generally speaking, consider the "entropy" an API client would experience.
+
+### Using one's own API
+
+The #1 best way to understand and address the weaknesses in an API's design and implementation is to use it in a production system.
+
+Whenever feasible, design an API in parallel with an accompanying integration of that API.
+
+### Point of contact
+
+Have an obvious mechanism for clients to report issues and ask questions about the API.
+
+When using GitHub for an API's code, use the associated issue tracker. In addition, publish an email address for direct, non-public inquiries.
+
+### Notifications of updates
+
+Have a simple mechanism for clients to follow changes to the API.
+
+Common ways to do this include a mailing list, or a [dedicated developer blog](https://developer.github.com/changes/) with an RSS feed.
+
+### API Endpoints
+
+An "endpoint" is a combination of two things:
+
+* The verb (e.g. `GET` or `POST`)
+* The URL path (e.g. `/articles`)
+
+Information can be passed to an endpoint in either of two ways:
+
+* The URL query string (e.g. `?year=2014`)
+* HTTP headers (e.g. `X-Api-Key: my-key`)
+
+When people say "RESTful" nowadays, they really mean designing simple, intuitive endpoints that represent unique functions in the API.
+
+Generally speaking:
+
+* **Avoid single-endpoint APIs.** Don't jam multiple operations into the same endpoint with the same HTTP verb.
+* **Prioritize simplicity.** It should be easy to guess what an endpoint does by looking at the URL and HTTP verb, without needing to see a query string.
+* Endpoint URLs should advertise resources, and **avoid verbs**.
```

Comment:
I think maybe some clarity is warranted here. I too dislike the blanket **avoid verbs**, but I don't think it's necessarily bad advice. It definitely needs more clarification than a full-stop though.

In the Sunlight example, `/search` and `/locate` should simply be query strings on the nouns. Instead of `/bills/search?query=...`, I would expect it to simply take a full-text `query` argument to `/bills` (which their API does for other means of searching bills).